### ISSUE TITLE
Change Request: make `createIndexMap` error out and log debug info if it would infinite loop due to invalid arguments

### ISSUE THREAD

**Issue by jedwards1211** [eslint/eslint#20492]
### ESLint version

v9.39.2

### What problem do you want to solve?

A lot of people have experienced pegged CPU usage with `typescript-eslint` and `vscode-eslint` but [no one has been able to come up with a repro](https://github.com/typescript-eslint/typescript-eslint/issues/1042).  One user was able to identify that `createIndexMap` was stuck in an infinite loop: https://github.com/microsoft/vscode-eslint/issues/721#issuecomment-526366654

Looking at the code in `createIndexMap`, there's no guarantee the loop will terminate, an invalid combination of `tokens` and `comments` could cause it to get stuck in an infinite loop.

### What do you think is the correct solution?

`createIndexMap` could easily detect if the loop is failing to make progress and error out:

```js
while (tokenIndex < tokens.length || commentIndex < comments.length) {
  const lastTokenIndex = tokenIndex, lastCommentIndex = commentIndex

  // ...

  if (tokenIndex === lastTokenIndex && commentIndex === lastCommentIndex) {
    // log other information?
    throw new Error(`invalid tokens or comments; unable to create an index map`)
  }
}
```

Ideally, we would also log which file is triggering the issue in whichever caller has that information.

This would at least prevent the CPU pegging and hopefully better surface the issue and provide information that helps people identify the root cause.

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by DMartens** [comment]
For this kind of change we need a reproduction as the whole mentioned discussion is over 5 years old for now unsupported versions of ESLint and `typescript-eslint`.
ESLint has certain expectations which the parser result must match (e.g. each node has a "type" property or that the tokens are sorted), otherwise we would not only need add a check here but also everywhere else in the code.
To get a reproduction you can patch the method as you describe and try to make it as minimal as possible.

**Comment by jedwards1211** [comment]
@DMartens you would rather have code you control potentially infinite loop when a plugin misbehaves, instead of erroring out?

If this is the issue I'm having I blame `typescript-eslint` for the invalid AST and I blame `eslint` for not catching it before it gets stuck in an infinite loop.

Spending hours trying to reproduce the issue in `typescript-eslint` feels like a waste of my time because even if I fix it, some other change in the future could accidentally cause `createIndexMap` to infinite loop again.  I would rather spend my time making the change described here, because not only will it help identify the issue in `typescript-eslint` whenever it happens to come up again (I have no idea which file is triggering it and there's no telling how long that would take to figure out, I have dozens of projects to install a patched version in), it will prevent similar issues in the future from eating my laptop battery.

Plus I would have thought that most people would think that loop looks unnecessarily risky, but maybe my standards are just too high.

**Comment by jedwards1211** [comment]
@DMartens  If I can make a test parser plugin that causes ESLint to get stuck in an infinite loop due to a malformed AST (that ESLint doesn't detect problems with before it gets to this point), will you accept that as a repro?

**Comment by DMartens** [comment]
I would only accept a reproduction which conforms to the [AST specification](https://eslint.org/docs/latest/extend/custom-parsers#ast-specification) as of course the loop can break when the produced AST / tokens do not conform.
Again my point is that given the conformance constraints (tokens are sorted and non-overlapping) the loop will stop and as such the condition would be unnecessary.
Further this is only my opinion, the rest of the team is welcome to weigh in (this is why I set the issue to "feedback needed").

**Comment by jedwards1211** [comment]
@DMartens why do you want to live in a world of insane software?

It's better for users if a program that accepts 3rd-party plugins does what it can to avoid totally insane behavior if plugins are buggy.  Infinite looping is an insane behavior.  You can't prevent a plugin's code from infinite looping, but you can prevent your own code in this case.

Good lord I think your opinion is terrible.
